### DISCUSSION TITLE
Gossipsub Topic Subscription Discovery

### DISCUSSION THREAD

**Discussion by yellowred** [libp2p/rust-libp2p#3740] [Q&A]
Hi,

I am trying to understand how subscription to topics in Gossipsub works so nodes can reliably send messages. In my setup, nodes A and B communicate via Gossipsub. Node A dials B and then sends a message, and B listens to that message. Both nodes use chat example to initialize swarm with mdns + I added identify, and then subscribe to a topic T on start up. If I start node B first, then, when A triggers an action that dials B, gossipsub at A does not see that B subscribed to T and unable to send a Message. However when I start A first and then B, when A connects to B, it is able to recognize that B is subscribed to T. 

I would expect that if a node dialed another node, then it would be able to see subscriptions, but somehow they are not available. How do I make it more reliable for nodes to stay connected and subscribed? My setup has small number of nodes that need to be able to quickly connect to each other to broadcast messages in a timely manner.

**Comment by mxinden** [comment]
That behavior is surprising to me. Thanks for documenting it here. I would expect each of the nodes [to send their subscriptions on a new connection](https://github.com/libp2p/rust-libp2p/blob/23c3d2461031b006f7fbccda0740790cdd5d3d2b/protocols/gossipsub/src/behaviour.rs#L3093-L3100).

Were you able to root cause this issue further? Can you provide a  small reproducer, ideally in the form of a unit test?
